crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -134,52 +127,49 @@ impl Display for DisplayUnionType<'_> {
             }
         }
 
-        let mut first = true;
-
-        // Print all types, but write all literals together (while preserving their position).
-        for ty in elements {
+        let elements = elements.iter().filter_map(|ty| {
             if let Ok(literal_kind) = LiteralTypeKind::try_from(*ty) {
-                let Some(mut literals) = grouped_literals.remove(&literal_kind) else {
-                    continue;
-                };
-
-                if !first {
-                    f.write_str(" | ")?;
-                };
-
-                f.write_str("Literal[")?;
-
+                let mut literals = grouped_literals.remove(&literal_kind)?;
                 if literal_kind == LiteralTypeKind::IntLiteral {
                     literals.sort_unstable_by_key(|ty| ty.expect_int_literal());
                 }
-
-                for (i, literal_ty) in literals.iter().enumerate() {
-                    if i > 0 {
-                        f.write_str(", ")?;
-                    }
-                    literal_ty.representation(self.db).fmt(f)?;
-                }
-                f.write_str("]")?;
+                Some(DisplayUnionElement::LiteralGroup(literals, self.db))
             } else {
-                if !first {
-                    f.write_str(" | ")?;
-                };
-
-                ty.display(self.db).fmt(f)?;
+                Some(DisplayUnionElement::SingleType(ty.display(self.db)))
             }
+        });
 
-            first = false;
-        }
+        f.join(" | ").entries(elements).finish()?;
```

Comment:
Took me a while to understand why we need `DisplayUnionElement`. I would probably rewrite this to using a `for` loop and calling `f.entry(&ty)` or `f.entry(&DisplayLiteralGroup(literals, self.db))`

Reply:
That makes sense. I think I kind of forgot about the `.entry` option when working on this, and assumed I needed to get things into an "iterator of items" shape. Oops.

I do like extracting the formatting of a literal group into its own type, so will probably keep that; it looks like that is your suggestion as well. Done and pushed.